hotrod-client-decoder/src/main/java/org/infinispan/hotrod/numeric/BranchlessParser.java

```diff
@@ -0,0 +1,120 @@
+      int wholeWithContinuations = wholeOrMore & thisVarintMask;
+      // mix them up as per varint spec while dropping the continuation bits:
+      // 0x7F007F isolate the first byte and the third byte dropping the continuation bits
+      // 0x7F007F00 isolate the second byte and the fourth byte dropping the continuation bits
+      // the second and fourth byte are shifted to the right by 1, filling the gaps left by the first and third byte
+      // it means that the first and second bytes now occupy the first 14 bits (7 bits each)
+      // and the third and fourth bytes occupy the next 14 bits (7 bits each), with a gap between the 2s of 2 bytes
+      // and another gap of 2 bytes after the forth and third.
+      wholeWithContinuations = (wholeWithContinuations & 0x7F007F) | ((wholeWithContinuations & 0x7F007F00) >> 1);
+      // 0x3FFF isolate the first 14 bits i.e. the first and second bytes
+      // 0x3FFF0000 isolate the next 14 bits i.e. the third and forth bytes
+      // the third and forth bytes are shifted to the right by 2, filling the gaps left by the first and second bytes
+      return (wholeWithContinuations & 0x3FFF) | ((wholeWithContinuations & 0x3FFF0000) >> 2);
+   }
+
+   private static int readRawVarInt40(ByteBuf buffer, int wholeOrMore) {
+      byte lastByte;
+      if (buffer.readableBytes() == 4 || (lastByte = buffer.getByte(buffer.readerIndex() + 4)) < 0) {
+         throw new CorruptedFrameException("malformed varint.");
+      }
+      buffer.skipBytes(5);
+      return (lastByte << 28) | (wholeOrMore & 0x3FFF) | ((wholeOrMore & 0x3FFF0000) >> 2);
+   }
+
+   public static long readRawVarint64(ByteBuf buffer) {
+      if (buffer.readableBytes() <= 4)
+         return readRawVarint32(buffer);
+
+      long wholeOrMore = buffer.getLongLE(buffer.readerIndex());
+      long firstOneOnStop = ~wholeOrMore & 0x8080808080808080L;
+
+      // The value occupies the 9 bytes. We just unroll it and consume the bytes.
+      if (firstOneOnStop == 0) {
+         return readRawVarInt72(buffer, wholeOrMore);
+      }
+
+      // Consume the bytes containing the long.
+      int bitsToKeep = Long.numberOfTrailingZeros(firstOneOnStop) + 1;
+      buffer.skipBytes(bitsToKeep >> 3);
+
+      // Create a mask and create the continuation bytes for decoding.
+      long thisVarlongMask = firstOneOnStop ^ (firstOneOnStop - 1);
+      long continuation = wholeOrMore & thisVarlongMask;
+
+      // We parse it as groups of bytes, first bytes 1, 3, 5, and 7.
+      // The second group is 2, 4, 6, and 8, which need a shift to right to compensate the gap.
+      continuation = (continuation & 0x007F007F007F007FL) | ((continuation & 0x7F007F007F007F00L) >> 1);
+
+      // Now we isolate the bits in sequence. We check 14 bits at a time.
+      // The intervals are 0-14 bits, 16-30 (and shift 2), 32-46 (and shift 2 + 2), 48-62 (and shift 2 + 2 + 2).
+      return (continuation & 0x3FFF) |
```

Comment:
you can use Long::compress if you have the right JDK version (since 19) using the right mask which isolates the bits you need "to compress"
see https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/lang/Long.html#compress(long,long)

Reply:
Thanks! We're on 17, but I'll add it as a comment to the ISPN code.